### ISSUE TITLE
data aws_ami cannot handle tf variables in parameters

### ISSUE THREAD

**Issue by gtmtech** [hashicorp/terraform#9945]
Terraform 0.7.9

This:

```
  filter {
    name     = "tag:workflow_${var.environment["name"]}"
    values   = ["published"]
  }
```

Results in this:

```
InvalidParameterValue: The filter '' is invalid
	status code: 400, request id: 17bcbba5-43e2-43bc-bd07-4759df25e
```

filters cannot handle basic terraform variables like just about everything else can

**Comment by apparentlymart** [comment]
Hi @gtmtech,

I can't think of any good reason why the handling of variables would be different for this particular resource, since the interpolation is (as always) handled by core Terraform rather than by the data source code.

So I think something else is going on here... would it be possible for you to share the output of running this with a `TF_LOG=debug` environment variable set (being careful to remove any sensitive info) via a gist? That would help me understand what exactly Terraform is sending to AWS to execute this query.

If you'd rather not share the entire debug output and are willing to dig in to the log a bit yourself, what I'm interested to see is where the request and response for the `DescribeImages` call are logged by the AWS client. If you search in the log output for `DescribeImages` it should hopefully be pretty obvious... there's usually a HTTP request and associated response just printed out into the log, almost verbatim.

**Comment by gtmtech** [comment]
Well this is very odd indeed. Two slightly different configs for the `aws_ami` resource provide entirely different results, despite having the same keys. 

This one produces the error consistently:

```
data "aws_ami" "citool" {
  most_recent = true
  executable_users = ["self"]
  filter {
    name = "tag:component"
    values = ["citool"]
  }
  filter {
    name = "tag:workflow_${environment["name"]}"
    values = ["published"]
  }
  owners = ["self"]
}
```

Which produces:

```
Error applying plan:

2 error(s) occurred:

* data.aws_ami.citool: InvalidParameterValue: The filter '' is invalid
	status code: 400, request id: 777c84b8-b152-4d31-8eea-96603a83bd3d
```

And with debug turned on:

```
2016/11/11 16:34:12 [ERROR] Shadow graph error: 3 error(s) occurred:

* ReadDataDiff "data.aws_ami.citool" had unequal configs (real, then shadow):

&terraform.ResourceConfig{ComputedKeys:[]string{"filter.1.name"}, Raw:map[string]interface {}{"executable_users":[]interface {}{"self"}, "filter":[]map[string]interface {}{map[string]interface {}{"name":"tag:component", "values":[]interface {}{"citool"}}, map[string]interface {}{"name":"tag:workflow_${environment[\"name\"]}", "values":[]interface {}{"published"}}}, "owners":[]interface {}{"self"}, "most_recent":true}, Config:map[string]interface {}{"executable_users":[]interface {}{"self"}, "filter":[]map[string]interface {}{map[string]interface {}{"name":"tag:component", "values":[]interface {}{"citool"}}, map[string]interface {}{"values":[]interface {}{"published"}}}, "owners":[]interface {}{"self"}, "most_recent":true}, raw:(*config.RawConfig)(0xc4215f29c0)}

&terraform.ResourceConfig{ComputedKeys:[]string{"filter.name"}, Raw:map[string]interface {}{"most_recent":true, "executable_users":[]interface {}{"self"}, "filter":[]map[string]interface {}{map[string]interface {}{"name":"tag:component", "values":[]interface {}{"citool"}}, map[string]interface {}{"values":[]interface {}{"published"}, "name":"tag:workflow_${environment[\"name\"]}"}}, "owners":[]interface {}{"self"}}, Config:map[string]interface {}{"filter":[]map[string]interface {}{map[string]interface {}{"name":"tag:component", "values":[]interface {}{"citool"}}, map[string]interface {}{"values":[]interface {}{"published"}}}, "owners":[]interface {}{"self"}, "most_recent":true, "executable_users":[]interface {}{"self"}}, raw:(*config.RawConfig)(0xc420e5f740)}
* data.aws_ami.citool: InvalidParameterValue: The filter '' is invalid
        status code: 400, request id: 777c84b8-b152-4d31-8eea-96603a83bd3d
```

Critically, there is no `DescribeImages` associated with this call, it doesnt ever seem to get that far. 


However, when I alter the resource to this:

```
data "aws_ami" "citool" {
  most_recent      = true
  executable_users = ["self"]
  owners           = ["self"]

  filter {
    name     = "tag:component"
    values   = ["citool"]
  }
  filter {
    name     = "tag:workflow_${var.environment["name"]}"
    values   = ["published"]
  }
}
```

Then it all works fine. 

Critically the difference above seems to be the `${var.environment["name"]}` vs `${environment["name"]}`, which was never going to exist. I think the original description was misleading, and the one in #9944 was the one causing the error, my apologies. 


So the bug can be summed up - if interpolation fails in the filter due to a lookup error, rather than keeping the non-interpolated part, it seems to drop the name entirely. It should perhaps have said:

```
filter "tag:workflow_" produces no results
```

or

```
error: ${environment["name"]} not found
```

**Comment by apparentlymart** [comment]
Ahh... This looks like yet another weird behavior Terraform exhibits when given an interpolation variable with no periods. Terraform 0.7 has been allowing this during validation due to a bug, but then doing all sorts of odd things when evaluating.

We have a fix for this coming in 0.8, though the fix will just be too make your original example raise an error, so at least you can see immediately that is isn't valid.

Thanks for the response. It seems like you have a working solution now, and we have a separate PR for the validation fix going into 0.8, so I'm going to close this... do let me know if I misunderstood and there's more to look at here.
